File: PE-common/src/nl/adaptivity/process/engine/PETransformer.java
package nl.adaptivity.process.engine;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import javax.xml.bind.JAXB;
import javax.xml.bind.JAXBElement;
import javax.xml.namespace.QName;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.stream.XMLEventFactory;
import javax.xml.stream.XMLEventReader;
import javax.xml.stream.XMLEventWriter;
import javax.xml.stream.XMLInputFactory;
import javax.xml.stream.XMLOutputFactory;
import javax.xml.stream.XMLStreamException;
import javax.xml.stream.events.Attribute;
import javax.xml.stream.events.Characters;
import javax.xml.stream.events.Namespace;
import javax.xml.stream.events.StartElement;
import javax.xml.stream.events.XMLEvent;
import javax.xml.transform.Result;
import javax.xml.transform.Source;
import javax.xml.transform.dom.DOMResult;
import javax.xml.transform.dom.DOMSource;

import nl.adaptivity.process.util.Constants;
import nl.adaptivity.util.activation.Sources;
import nl.adaptivity.util.xml.AbstractBufferedEventReader;
import nl.adaptivity.util.xml.NodeEventReader;
import nl.adaptivity.util.xml.XmlUtil;

import org.w3c.dom.Document;
import org.w3c.dom.DocumentFragment;
import org.w3c.dom.Node;


public class PETransformer {





  public static class MyFilter extends AbstractBufferedEventReader {

    private PETransformerContext aContext;
    private XMLEventReader aInput;
    private XMLEventFactory aXef;

    public MyFilter(PETransformerContext pContext, XMLEventReader pInput) {
      aContext = pContext;
      aInput = pInput;
      aXef = XMLEventFactory.newInstance();
    }

    @Override
    public void remove() {
      throw new UnsupportedOperationException("Deletion is not supported on this read-only filter");
    }



    @Override
    public XMLEvent peek() throws XMLStreamException {
      if (! isPeekBufferEmpty()) {
        return peekFirst();
      }
      while(aInput.hasNext()) {
        XMLEvent event = aInput.nextEvent();
        if (event.isStartElement()) {
          return peekStartElement(event.asStartElement());
        } else if (event.isNamespace()){
          if (! Constants.MODIFY_NS.equals(((Namespace)event).getNamespaceURI())) {
            add(event);
            return event;
          }
        } else if (event.isCharacters()) {
          if (! isIgnorableWhiteSpace(event.asCharacters())) {
            add(event);
            return event;
          }
        } else {
          add(event);
          return event;
        }
      }
      return null;
    }

    private static boolean isIgnorableWhiteSpace(Characters pCharacters) {
      if (pCharacters.isIgnorableWhiteSpace()) {
        return true;
      }
      return XmlUtil.isXmlWhitespace(pCharacters.getData());
    }

    private XMLEvent peekStartElement(StartElement pElement) throws XMLStreamException {
      if (Constants.MODIFY_NS.toString().equals(pElement.getName().getNamespaceURI())) {
        String localname = pElement.getName().getLocalPart();

        final Map<String, String> attributes = parseAttributes(aInput, pElement);

        switch (localname) {
          case "attribute":
            add(getAttribute(attributes));
            readEndTag(pElement.getName());
            return peekFirst();
          case "element":
            processElement(attributes, false);
            readEndTag(pElement.getName());
            return peekFirst();
          case "value":
            processElement(attributes, true);
            readEndTag(pElement.getName());
            return peekFirst();
          default:
            throw new XMLStreamException("Unsupported element: "+pElement.getName());
        }
      } else {
        add(pElement);
        return pElement;
      }
    }

    private void readEndTag(QName pName) throws XMLStreamException {
      XMLEvent ev = aInput.nextTag();
      if (! (ev.isEndElement() && ev.asEndElement().getName().equals(pName))) {
        throw new XMLStreamException("Unexpected tag found ("+ev+")when expecting an end tag for "+pName);
      }
    }

    private void processElement(Map<String,String> pAttributes, boolean pHasDefault) throws XMLStreamException {
      String valueName = pAttributes.get("value");
      if (valueName==null) {
        if (pHasDefault) {
          addAll(aContext.resolveDefaultValue(aXef));
        } else {
          throw new XMLStreamException("This context does not allow for a missing value parameter");
        }
      } else {
        addAll(aContext.resolveElementValue(aXef, valueName));
      }
    }

    private static Map<String,String> parseAttributes(XMLEventReader pIn, StartElement pStartElement) throws XMLStreamException {
      TreeMap<String, String> result = new TreeMap<>();

      @SuppressWarnings("unchecked")
      Iterator<Attribute> attributes = pStartElement.getAttributes();
      while (attributes.hasNext()) {
        Attribute attribute = attributes.next();
        result.put(attribute.getName().getLocalPart(), attribute.getValue());
      }

      while(pIn.peek().isAttribute()) {
        Attribute attribute = (Attribute) pIn.nextEvent();
        result.put(attribute.getName().getLocalPart(), attribute.getValue());
      }
      return result;
    }

    private XMLEvent getAttribute(Map<String,String> pAttributes) throws XMLStreamException {
      String valueName = pAttributes.get("value");
      String paramName = pAttributes.get("name");

      if (valueName != null) {
        if (paramName==null) {
          paramName = aContext.resolveAttributeName(valueName);
        }
        String value = aContext.resolveAttributeValue(valueName);
        return aXef.createAttribute(paramName, value);
      } else {
        throw new MessagingFormatException("Missing parameter name");
      }
    }

    @Override
    public Object getProperty(String pName) throws IllegalArgumentException {
      return aInput.getProperty(pName);
    }

    @Override
    public void close() throws XMLStreamException {
      aInput.close();
      aInput = null;
      aContext = null;
      aXef = null;
      super.close();
    }

  }

  public interface PETransformerContext {

    List<XMLEvent> resolveElementValue(XMLEventFactory pXef, String pValueName) throws XMLStreamException;
    List<XMLEvent> resolveDefaultValue(XMLEventFactory pXef) throws XMLStreamException;
    String resolveAttributeValue(String pValueName) throws XMLStreamException;
    String resolveAttributeName(String pValueName);

  }

  public static abstract class AbstractDataContext implements PETransformerContext {

    protected abstract ProcessData getData(String pValueName);

    @Override
    public List<XMLEvent> resolveElementValue(XMLEventFactory pXef, String pValueName) throws XMLStreamException {
      ProcessData data = getData(pValueName);
      if (data==null) {
        throw new IllegalArgumentException("No value with name "+pValueName+" found");
      }
      return toEvents(data);
    }

    protected List<XMLEvent> toEvents(ProcessData data) throws XMLStreamException {
      List<XMLEvent> result = new ArrayList<>();
      DocumentFragment frag = data.getDocumentFragment();
      for(XMLEventReader dataReader = new NodeEventReader(frag);dataReader.hasNext();) {
        result.add(dataReader.nextEvent());
      }
      return result;
    }

    @Override
    public String resolveAttributeValue(String pValueName) throws XMLStreamException {
      ProcessData data = getData(pValueName);
      XMLEventReader dataReader = new NodeEventReader(data.getDocumentFragment());
      StringBuilder result = new StringBuilder();
      while (dataReader.hasNext()) {
        XMLEvent event = dataReader.nextEvent();
        switch (event.getEventType()) {
        case XMLEvent.ATTRIBUTE:
        case XMLEvent.DTD:
        case XMLEvent.START_ELEMENT:
        case XMLEvent.END_ELEMENT:
          throw new XMLStreamException("Unexpected node found while resolving attribute. Only CDATA allowed: "+event);
        case XMLEvent.CDATA:
        case XMLEvent.CHARACTERS:
          result.append(event.asCharacters().getData());
          break;
        case XMLEvent.START_DOCUMENT:
        case XMLEvent.END_DOCUMENT:
        case XMLEvent.COMMENT:
        case XMLEvent.PROCESSING_INSTRUCTION:
          break; // ignore
        default:
          throw new XMLStreamException("Unexpected node type: "+event);
        }
      }
      return result.toString();
    }

    @Override
    public String resolveAttributeName(String pValueName) {
      ProcessData data = getData(pValueName);
      return XmlUtil.toString(data.getDocumentFragment());
    }

  }

  public static class ProcessDataContext extends AbstractDataContext {

    private ProcessData[] aProcessData;
    private int aDefaultIdx;

    public ProcessDataContext(ProcessData... pProcessData) {
      aProcessData = pProcessData;
      aDefaultIdx = pProcessData.length==1 ? 0 : -1;
    }

    public ProcessDataContext(int pDefaultIdx, ProcessData... pProcessData) {
      assert pDefaultIdx>=-1 && pDefaultIdx<pProcessData.length;
      aProcessData = pProcessData;
      aDefaultIdx = pDefaultIdx;
    }

    @Override
    protected ProcessData getData(String pValueName) {
      for(ProcessData candidate: aProcessData) {
        if (pValueName.equals(candidate)) { return candidate; }
      }
      return null;
    }

    @Override
    public List<XMLEvent> resolveDefaultValue(XMLEventFactory pXef) throws XMLStreamException {
      return toEvents(aProcessData[aDefaultIdx]);
    }

  }

  private final PETransformerContext aContext;

  private PETransformer(PETransformerContext pContext) {
    aContext = pContext;
  }

  public static PETransformer create(ProcessData... pProcessData) {
    return new PETransformer(new ProcessDataContext(pProcessData));
  }

  public static PETransformer create(PETransformerContext pContext) {
    return new PETransformer(pContext);
  }

  public List<Node> transform(List<? extends Object> pContent) {
    try {
      Document document = null;
      ArrayList<Node> result = new ArrayList<>(pContent.size());
      for(Object obj: pContent) {
        if (obj instanceof CharSequence) {
          if (document == null) {
            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
            dbf.setNamespaceAware(true);
            document = dbf.newDocumentBuilder().newDocument();
          }
          result.add(document.createTextNode(obj.toString()));
        } else if (obj instanceof Node) {
          if (document==null) { document = ((Node) obj).getOwnerDocument(); }
          DocumentFragment v = transform((Node) obj);
          if (v!=null) {
            result.add(v);
          }
        } else if (obj instanceof JAXBElement<?>) {
          if (document == null) {
            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
            dbf.setNamespaceAware(true);
            document = dbf.newDocumentBuilder().newDocument();
          }
          JAXBElement<?> jbe = (JAXBElement<?>) obj;
          final DocumentFragment df = document.createDocumentFragment();
          DOMResult domResult = new DOMResult(df);
          JAXB.marshal(jbe, domResult);
          for(Node n = df.getFirstChild(); n!=null; n=n.getNextSibling()) {
            DocumentFragment v = transform(n);
            if (v!=null) {
              result.add(v);
            }
          }
        } else if (obj!=null) {
          throw new IllegalArgumentException("The node "+obj.toString()+" of type "+obj.getClass()+" is not understood");
        }
      }
      return result;
    } catch (ParserConfigurationException e) {
      throw new RuntimeException(e);
    }
  }

  public DocumentFragment transform(Node pNode) {
    DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
    Document document;
    try {
      document = dbf.newDocumentBuilder().newDocument();
      DocumentFragment fragment = document.createDocumentFragment();
      DOMResult result = new DOMResult(fragment);
      transform(new DOMSource(pNode), result);
      return fragment;
    } catch (ParserConfigurationException | XMLStreamException e) {
      throw new RuntimeException(e);
    }
  }

  public void transform(Source source, final Result result) throws XMLStreamException {
    final XMLInputFactory xif = XMLInputFactory.newInstance();
    final XMLOutputFactory xof = XMLOutputFactory.newInstance();
    final XMLEventReader xer = createFilter(xif.createXMLEventReader(Sources.toReader(source)));
    final XMLEventWriter xew = xof.createXMLEventWriter(result);
    xew.add(xer);
  }

  private XMLEventReader createFilter(XMLEventReader pInput) {
    return new MyFilter(aContext, pInput);
  }

}

File: ProcessEngine/core/src/commonMain/kotlin/nl/adaptivity/process/engine/PETransformer.kt
/*
 * Copyright (c) 2019.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.engine

import nl.adaptivity.process.engine.impl.dom.*
import nl.adaptivity.process.engine.impl.getClass
import nl.adaptivity.process.util.Constants
import nl.adaptivity.util.multiplatform.assert
import nl.adaptivity.util.xml.CombinedNamespaceContext
import nl.adaptivity.xmlutil.*
import nl.adaptivity.xmlutil.XmlEvent.*
import kotlin.jvm.JvmOverloads


class PETransformer private constructor(
    private val context: PETransformerContext,
    private val namespaceContext: NamespaceContext?,
    private val isRemoveWhitespace: Boolean
) {

    private class TransformingFilter(
        private val context: PETransformerContext,
        private val filterNamespaceContext: NamespaceContext?,
        delegate: XmlReader,
        private val isRemoveWhitespace: Boolean
    ) : XmlBufferedReader(delegate) {

        @OptIn(XmlUtilInternal::class)
        public override fun doPeek(): List<XmlEvent> {
            val results = ArrayList<XmlEvent>(1)

            doPeek(results)
            return results
        }

        @OptIn(XmlUtilInternal::class)
        private fun doPeek(results: MutableList<XmlEvent>) {
            val events = super.doPeek()

            loop@ for (event in events) {
                when (event.eventType) {
                    EventType.START_ELEMENT        -> peekStartElement(results, event as StartElementEvent)
                    EventType.TEXT                 -> {
                        val text = event as TextEvent
                        if (!isIgnorableWhiteSpace(text)) {
                            results.add(event)
                            continue@loop
                        }
                        // fall through if ignorable
                        if (!isRemoveWhitespace) {
                            results.add(event)
                        }
                        doPeek(results) //peek again, as sometimes whitespace needs to be stripped to add attributes
                        // fall through if not whitespace
                    }
                    EventType.IGNORABLE_WHITESPACE -> {
                        if (!isRemoveWhitespace) {
                            results.add(event)
                        }
                        doPeek(results)
                    }
                    else                           -> results.add(event)
                }
            }
        }


        private fun stripWhiteSpaceFromPeekBuffer(results: MutableList<XmlEvent>) {
            while (results.size > 0 && ((results.last() as? TextEvent)?.let { isXmlWhitespace(it.text) } == true)) {
                results.removeAt(results.size - 1)
            }
        }

        private fun peekStartElement(results: MutableList<XmlEvent>, element: StartElementEvent) {
            if (Constants.MODIFY_NS_STR == element.namespaceUri) {
                val localname = element.localName

                val attributes = parseAttributes(element)

                when (localname) {
                    "attribute" -> {
                        stripWhiteSpaceFromPeekBuffer(results)
                        results.add(getAttribute(attributes))
                        readEndTag(element)
                        return
                    }
                    "element"   -> {
                        processElement(results, element, attributes, false)
                        readEndTag(element)
                        return
                    }
                    "value"     -> {
                        processElement(results, element, attributes, true)
                        readEndTag(element)
                        return
                    }
                    else        -> throw XmlException("Unsupported element: {${element.namespaceUri}}${element.localName}")
                }
            } else {
                var filterAttributes = false
                val newAttrs = ArrayList<Attribute>()
                for (attr in element.attributes) {
                    if (attr.hasNamespaceUri() && Constants.MODIFY_NS_STR == attr.value) {
                        filterAttributes = true
                    } else {
                        newAttrs.add(attr)
                    }
                }
                val newNamespaces = ArrayList<Namespace>()
                for (ns in element.namespaceDecls) {
                    if (Constants.MODIFY_NS_STR == ns.namespaceURI) {
                        filterAttributes = true
                    } else {
                        newNamespaces.add(ns)
                    }
                }
                if (filterAttributes) {
                    results.add(
                        StartElementEvent(
                            element.locationInfo, element.namespaceUri, element.localName,
                            element.prefix,
                            newAttrs.toTypedArray(),
                            namespaceContext.freeze(),
                            newNamespaces
                        )
                    )
                } else {
                    results.add(element)
                }
            }
        }

        @OptIn(XmlUtilInternal::class)
        private fun readEndTag(name: StartElementEvent) {
            while (true) {
                val elems = super.doPeek()
                loop@ for (elem in elems) {
                    when (elem.eventType) {
                        EventType.IGNORABLE_WHITESPACE, EventType.COMMENT -> {
                        }
                        EventType.TEXT                                    -> {
                            if (isXmlWhitespace((elem as TextEvent).text)) {
                                continue@loop
                            }
                            if (!(elem.eventType === EventType.END_ELEMENT && name.isEqualNames(
                                    elem as EndElementEvent
                                ))
                            ) {
                                throw XmlException(
                                    "Unexpected tag found ($elem) when expecting an end tag for $name"
                                )
                            }
                            return
                        }
                        else                                              -> {
                            if (!(elem.eventType === EventType.END_ELEMENT && name.isEqualNames(
                                    elem as EndElementEvent
                                ))
                            ) {
                                throw XmlException(
                                    "Unexpected tag found ($elem) when expecting an end tag for $name"
                                )
                            }
                            return
                        }
                    }
                }
            }
        }

        private fun processElement(
            results: MutableList<XmlEvent>,
            event: StartElementEvent,
            attributes: Map<String, CharSequence>,
            hasDefault: Boolean
        ) {
            val valueName = attributes["value"]
            val xpath = attributes["xpath"]
            if (valueName == null) {
                if (hasDefault) {
                    addAllRegular(
                        results,
                        applyXpath(event.namespaceContext, context.resolveDefaultValue(), xpath)
                    )
                } else {
                    throw XmlException("This context does not allow for a missing value parameter")
                }
            } else {
                addAllRegular(
                    results,
                    applyXpath(event.namespaceContext, context.resolveElementValue(valueName), xpath)
                )
            }
        }

        private fun addAllRegular(target: MutableList<XmlEvent>, source: Iterable<XmlEvent>) {
            for (event in source) {
                if (!event.isIgnorable) {
                    target.add(event)
                }
            }
        }

        @OptIn(XmlUtilInternal::class)
        private fun applyXpath(
            namespaceContext: NamespaceContext,
            pendingEvents: List<XmlEvent>,
            xpath: CharSequence?
        ): Collection<XmlEvent> {
            val xpathstr = xpath?.toString()
            if (xpathstr == null || "." == xpathstr) {
                return pendingEvents
            }
            // TODO add a function resolver
            val rawPath = newXPath()
            // Do this better
            if (filterNamespaceContext == null) {
                rawPath.setNamespaceContext(namespaceContext)
            } else {
                rawPath.setNamespaceContext(CombinedNamespaceContext(namespaceContext, filterNamespaceContext))
            }
            val xpathexpr = rawPath.compile(xpathstr)
            val result = ArrayList<XmlEvent>()
            val dbf = newDocumentBuilderFactory()
            dbf.isNamespaceAware = true
            val db = dbf.newDocumentBuilder()
            val eventFragment = db.newDocument().createDocumentFragment()
            val domResult = DOMResult(eventFragment)

            val writer = domResult.newWriter()
            for (event in pendingEvents) {
                event.writeTo(writer)
            }
            writer.close()
            val applicationResult = xpathexpr.evaluate(eventFragment, XPathConstants.NODESET) as NodeList
            if (applicationResult.getLength() > 0) {
                result.addAll(toEvents(ProcessData("--xpath result--", applicationResult.toFragment())))
            }
            return result
        }

        private fun parseAttributes(startElement: StartElementEvent): Map<String, CharSequence> {
            val result = mutableMapOf<String, CharSequence>()

            for (attribute in startElement.attributes) {
                result[attribute.localName] = attribute.value
            }
            return result
        }

        private fun getAttribute(attributes: Map<String, CharSequence>): XmlEvent {
            val valueName = attributes["value"]?.toString()
            val xpath = attributes["xpath"]
            var paramName: CharSequence? = attributes["name"]

            if (valueName != null) {
                if (paramName == null) {
                    paramName = context.resolveAttributeName(valueName)
                }
                val value = context.resolveAttributeValue(valueName, xpath?.toString())
                return Attribute(
                    null, XMLConstants.NULL_NS_URI, paramName, XMLConstants.DEFAULT_NS_PREFIX,
                    value
                )
            } else {
                throw MessagingFormatException("Missing parameter name")
            }
        }
    }


    interface PETransformerContext {
        fun resolveElementValue(valueName: CharSequence): List<XmlEvent>
        fun resolveDefaultValue(): List<XmlEvent>  //Just return DOM, not events (that then need to be dom-ified)
        fun resolveAttributeValue(valueName: String, xpath: String?): String
        fun resolveAttributeName(valueName: String): String

    }

    abstract class AbstractDataContext : PETransformerContext {

        protected abstract fun getData(valueName: String): ProcessData?

        override fun resolveElementValue(valueName: CharSequence): List<XmlEvent> {
            val data = getData(valueName.toString()) ?: throw IllegalArgumentException(
                "No value with name $valueName found"
            )
            return toEvents(data)
        }

        override fun resolveAttributeValue(valueName: String, xpath: String?): String {
            val data = getData(valueName) ?: throw IllegalArgumentException("No data value with name $valueName found")
            val dataReader = data.contentStream
            val result = StringBuilder()
            try {
                while (dataReader.hasNext()) {
                    val event = dataReader.next()
                    when (event) {
                        EventType.ATTRIBUTE, EventType.DOCDECL, EventType.START_ELEMENT                                       -> throw XmlException(
                            "Unexpected node found while resolving attribute. Only CDATA allowed: (${event.getClass()}) " + event
                        )
                        EventType.CDSECT, EventType.TEXT                                                                      -> {
                            if (!isIgnorableWhiteSpace(dataReader)) {
                                result.append(dataReader.text)
                            }
                        }
                        EventType.START_DOCUMENT, EventType.END_DOCUMENT, EventType.COMMENT, EventType.PROCESSING_INSTRUCTION -> {
                        }
                        EventType.END_ELEMENT // finished the element
                                                                                                                              -> return result.toString()
                        else                                                                                                  -> throw XmlException(
                            "Unexpected node type: $event"
                        )
                    }// ignore
                }
            } catch (e: XmlException) {
                throw XmlException("Failure to parse data (name=$valueName, value=${data.content.contentString})", e)
            }

            return result.toString()
        }

        override fun resolveAttributeName(valueName: String): String {
            val data = getData(valueName)!!
            return data.content.content.concatToString()
        }

    }

    class ProcessDataContext : AbstractDataContext {

        private var processData: Array<ProcessData>
        private var defaultIdx: Int = 0

        constructor(vararg processData: ProcessData) {
            this.processData = Array(processData.size) { processData[it] }
            defaultIdx = if (processData.size == 1) 0 else -1
        }

        constructor(defaultIdx: Int, vararg processData: ProcessData) {
            assert(defaultIdx >= -1 && defaultIdx < processData.size)
            this.processData = Array(processData.size) { processData[it] }
            this.defaultIdx = defaultIdx
        }

        protected override fun getData(valueName: String): ProcessData? {
            for (candidate in processData) {
                if (valueName == candidate.name) {
                    return candidate
                }
            }
            return null
        }

        override fun resolveDefaultValue(): List<XmlEvent> {
            return when {
                processData.isEmpty() -> emptyList()
                else                  -> toEvents(processData[defaultIdx])
            }
        }

    }

    fun transform(input: XmlReader, out: XmlWriter) {
        val filteredIn = createFilter(input)
        while (filteredIn.hasNext()) {
            filteredIn.next() // Don't forget to move to next element as well.
            filteredIn.writeCurrent(out)
        }
    }

    fun createFilter(input: XmlReader): XmlReader {
        return TransformingFilter(context, namespaceContext, input, isRemoveWhitespace)
    }

    companion object {

        fun create(
            namespaceContext: NamespaceContext?,
            removeWhitespace: Boolean,
            vararg processData: ProcessData
        ): PETransformer {
            return PETransformer(ProcessDataContext(*processData), namespaceContext, removeWhitespace)
        }

        @Deprecated("")
        fun create(removeWhitespace: Boolean, vararg processData: ProcessData): PETransformer {
            return create(null, removeWhitespace, *processData)
        }

        fun create(namespaceContext: NamespaceContext, vararg processData: ProcessData): PETransformer {
            return create(namespaceContext, true, *processData)
        }

        @Deprecated("")
        fun create(vararg processData: ProcessData): PETransformer {
            return create(null, true, *processData)
        }

        @Deprecated("")
        fun create(context: PETransformerContext): PETransformer {
            return create(context, true)
        }

        fun create(context: PETransformerContext, removeWhitespace: Boolean): PETransformer {
            return create(null, context, removeWhitespace)
        }

        @JvmOverloads
        fun create(
            namespaceContext: NamespaceContext?,
            context: PETransformerContext,
            removeWhitespace: Boolean = true
        ): PETransformer {
            return PETransformer(context, namespaceContext, removeWhitespace)
        }

        protected fun toEvents(data: ProcessData): List<XmlEvent> {
            val result = ArrayList<XmlEvent>()

            val frag = data.contentStream
            while (frag.hasNext()) {
                frag.next()
                result.add(XmlEvent.from(frag))
            }
            return result
        }

        internal fun isIgnorableWhiteSpace(characters: TextEvent): Boolean {
            return if (characters.eventType === EventType.IGNORABLE_WHITESPACE) {
                true
            } else isXmlWhitespace(characters.text)
        }

        internal fun isIgnorableWhiteSpace(characters: XmlReader): Boolean {
            return if (characters.eventType === EventType.IGNORABLE_WHITESPACE) {
                true
            } else isXmlWhitespace(characters.text)
        }
    }
}
